### PR TITLE
[2/3] upgrade react native 0.71

### PR DESCRIPTION
**Pull request by Kudo**
# Why

follow up #20799 for react-native 0.71 upgrade. this pr aims for jest upgrade and fix all broken tests.
close ENG-7192

# How

- upgrade packages
  - `jest: ^26.0.24 -> ^29.2.1`
  - `jest-watch-typeahead: 0.6.4 -> 2.2.1`
  - `@types/jest: ^26.0.24 -> ^29.2.1`
  - `babel-jest: ^26.6.3 -> ^29.2.1`
  - `@jest/create-cache-key-function: ^27.0.1 -> ^29.2.1`
  - `@babel/preset-env: ^7.12.9 -> ^7.14.0`
  - `eslint: ^8.20.0 -> ^8.29.0`
  - add `jest-environment-jsdom` because new jest does not include it by default
- [core] fix `SyntheticPlatformEmitter` type error. originally it refers to `react-native/Libraries/EventEmitter/RCTDeviceEventEmitter` internal file which has incorrect type setup in 0.71. we could use the `react-native.DeviceEventEmitter` instead.
- [expo-linking]: remove the deprecated `Linking.removeEventListener` (which is also removed from upstream)
- [firebase-recaptcha] remove broken web test because firebase ESM which is not supported by jest. this pr simply remove the test case because the package is deprecated.
- [jest-expo-enzyme] remove this package and move to [the archived repo](https://github.com/expo/jest-expo-enzyme). enzyme cannot upgrade to jest 29.
- remove `@types/react-native` and the versioned cli package check. because 0.71 ships the types directly.
- [ncl] workaround `@react-native-community/slider`, `@react-native-segmented-control/segmented-control`, and `@react-native-masked-view/masked-view` type errors in the `react-native-71-fix.d.ts`.
- for other details, please check the commit histories one by one.

# Note

- ~will update changelogs later, so we could get first round reviews.~ update in #20858

# Test Plan

- ci passed
- for test-suite ios ci error, i've mentioned in #20799

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: .github/workflows/test-suite.yml [hunk 1]
```diff
@@ -49,6 +49,8 @@ jobs:
       - name: 🧪 Run Web tests
         run: yarn test:web
         working-directory: apps/bare-expo
+        env:
+          EXPO_NO_TYPESCRIPT_SETUP: 1
```
Comment (Kudo) [inline .github/workflows/test-suite.yml:53]: the global cli still want to install `@types/react-native`. before upgrade global cli we should keep this.

File: packages/@expo/cli/e2e/__tests__/export-web-test.ts [hunk 2]
```diff
@@ -26,8 +27,8 @@ it('loads expected modules by default', async () => {
     `require('../../build/src/export/web').expoExportWeb`
   );
   expect(modules).toStrictEqual([
-    '../node_modules/ansi-styles/index.js',
     '../node_modules/arg/index.js',
+    '../node_modules/chalk/node_modules/ansi-styles/index.js',
```
Comment (Kudo) [inline packages/@expo/cli/e2e/__tests__/export-web-test.ts:31]: the `ansi-styles` package is not hoised in the current monorepo situation.

File: packages/@expo/cli/src/start/doctor/typescript/TypeScriptProjectPrerequisite.ts [hunk 3]
```diff
@@ -77,7 +77,6 @@ export class TypeScriptProjectPrerequisite extends ProjectPrerequisite {
           // the package and attempts to resolve the module in the same process.
           { file: 'typescript/package.json', pkg: 'typescript' },
           { file: '@types/react/package.json', pkg: '@types/react' },
-          { file: '@types/react-native/package.json', pkg: '@types/react-native' },
```
Comment (Kudo) [inline packages/@expo/cli/src/start/doctor/typescript/TypeScriptProjectPrerequisite.ts:80]: note here i've removed the `@types/react-native` package check in versioned cli.

**Review by lukmccall — APPROVED** [review]
Greate work 🔥

**Review by tsapeta — APPROVED** [review]

**Comment by Kudo** [comment]
i'm going to merge this pr. if there's any feedback from post-review, please let me know, and i'll follow up separately.

**Pull request merged**